src/frontend/src/optimizer/plan_node/stream_group_topn.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use risingwave_common::util::sort_util::OrderType;
+
+use super::utils::TableCatalogBuilder;
+use super::PlanBase;
+use crate::optimizer::property::{Distribution, Order};
+use crate::{PlanRef, TableCatalog};
+
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct StreamGroupTopN {
+    pub base: PlanBase,
+    group_key: Vec<usize>,
+    limit: usize,
+    offset: usize,
+    order: Order,
+}
+
+#[allow(dead_code)]
+impl StreamGroupTopN {
+    pub fn new(
+        input: PlanRef,
+        group_key: Vec<usize>,
+        limit: usize,
+        offset: usize,
+        order: Order,
+    ) -> Self {
+        let base = PlanBase::new_stream(
+            input.ctx(),
+            input.schema().clone(),
+            input.logical_pk().to_vec(),
+            Distribution::HashShard(group_key.clone()),
```

Comment:
maybe it should be derived from the input, like HashShard -> HashShard, UpStreamHashShard->UpStreamHashShard, Single->Single